### PR TITLE
#88 Notifications

### LINKED ISSUE #77: Task 2.4: Set up notification
| Task | P | 
|:-----------------------:|:---:|
| 2.4: Set up notification| Medium |

When a user posts a comment, notification should be sent to the user who posted the picture.

The priority point of this task is medium because it is not part of the core feature, however, it is implied in the feature's description.

## Related features/user stories

* Related to [User Story 2](#13)

### LINKED ISSUE #88: Task 3.1 Set up notification
| Task| P |
|:-----------------------:|:---:|
|Set up notification  | Medium |

The user receives a notification when someone decides to follow them. This task has a Medium priority because it is not essential to the feature, but it is implied in the feature's description. 

## Related features/user stories

* Related to [User Story 3](#15)

### LINKED ISSUE #90: Task 3.2: Create a list of followers/people we follow
| Task| P |
|:-----------------------:|:---:|
|Create a list of followers/people we follow  | High |

A list of users that user A follows and a list of users that follow user A. This task is of High priority because we need it in order to pass the acceptance and to verify if the feature of following others even works as intended.

## Related features/user stories

* Related to [User Story 3](#15)

### LINKED ISSUE #91: Task 3.3: Set Acceptance tests
| Task| P |
|:-----------------------:|:---:|
|Set Acceptance Tests| Low |

There are a few acceptance tests that have been set for our user story on [Follow other](#15). This is a low priority since work on the feature has not yet begun. They are the following: 

------------------------------------------

## Acceptance Tests

| Input| Expected Output |
|:-----------------------:|:---:|
| Clicks on "follow" button of another user| The follow button changes to "unfollow" and the user is now following this other user, and generates a notification to the user that has been followed |

### LINKED ISSUE #92: Task 3.4: Pass Acceptance Tests
| Task| P |
|:-----------------------:|:---:|
|Pass Acceptance Tests| Low |

There are a few acceptance tests that have been set for our user story on [Follow other](#15). They must all pass with any submit. Priority is low because the work on the feature hasn't started yet. They are the following **(please reference a commit if you check an item off of the list):** 

------------------------------------------

## Acceptance Tests
- [x] List of followers/people we follow
- [x] Set up notification

### PR DESCRIPTION
**Pull request by tjosepo**
## Summary

This PR adds the Follow feature, and it also adds a notification icon that, once clicked, displays a few notifications.

## Related Issue

* Resolves #88 
* Resolves #90 
* Resolves #91 
* Resolves #92 
* Resolves #77

## Testing 

1. Log in. 
2. Follow someone (it can be yourself).
3. That person should receive a notification about being followed.
4. When that person posts a picture, its followers will be notified.

### CODE REVIEW

File: functions/tslint.json [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    "ban-comma-operator": true,
+
+    // Do not allow internal modules or namespaces . These are deprecated in favor of ES6 modules.
+    "no-namespace": true,
+
+    // Do not allow parameters to be reassigned. To avoid bugs, developers should instead assign new values to new vars.
+    "no-parameter-reassignment": true,
+
+    // Force the use of ES6-style imports instead of /// <reference path=> imports.
+    "no-reference": true,
```
Comment (laurent-lao) [inline functions/tslint.json:19]: I like the new rules

File: package.json [hunk 2]
```diff
@@ -46,11 +51,6 @@
     "protractor": "~5.4.3",
     "ts-node": "~8.3.0",
     "tslint": "~5.18.0",
-    "typescript": "~3.7.5",
-    "@angular-devkit/architect": "^0.900.0-0 || ^0.900.0",
-    "firebase-tools": "^7.12.1",
-    "fuzzy": "^0.1.3",
-    "inquirer": "^6.2.2",
-    "inquirer-autocomplete-prompt": "^1.0.1"
```
Comment (laurent-lao) [inline package.json:54]: Any reason why this was removed?

File: src/app/login-button/login-button.component.ts [hunk 3]
```diff
@@ -8,15 +8,10 @@ import { AuthService } from '../core/auth.service';
...
   constructor(public auth: AuthService, private router: Router) { }
 
-  ngOnInit(): void {
-    const loginButton = document.getElementById('login_button');
-    loginButton.style.display = 'block';
-  }
-
   onLoginClick() {
     this.router.navigateByUrl('/user-profile');
   }
```
Comment (laurent-lao) [inline src/app/login-button/login-button.component.ts]: @A-Robitaille was assigned to fix this

**Review by laurent-lao — APPROVED** [review]
Looks good to me, will have some pair programming with you so we can model some of your functions

File: package.json (hunk 2 above)
Comment (CamilBouzidi) [reply package.json:54]: A lot of them were moved up or down. None of them were actually removed!

File: src/app/app.module.ts [hunk 4]
```diff
@@ -17,6 +18,9 @@ import { MatInputModule } from '@angular/material/input';
 import { MatFormFieldModule } from '@angular/material/form-field';
 import { MatSnackBarModule } from '@angular/material/snack-bar';
 import { MatCardModule } from '@angular/material/card';
+import { MatProgressSpinnerModule } from '@angular/material/progress-spinner';
```
Comment (CamilBouzidi) [inline src/app/app.module.ts:21]: Small touches like these will set our project apart. Great idea!

File: src/app/follow-button/follow-button.component.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    });
+  }
+
+  follow = () => {
+    this.disable = true;
+    this.followCallable({
+      uid: this.uid
+    }).subscribe(() => {
+      this.disable = false;
+      this.cdRef.detectChanges(); // Without this, the disable sometimes fails to get updated.
```
Comment (CamilBouzidi) [inline src/app/follow-button/follow-button.component.ts:39]: do you know how often the changes are detected or how long it takes to detect them? We could maybe improve the performance of the website through that.

**Review by CamilBouzidi — APPROVED** [review]
I left a comment or two for additional improvements, but they would not have to be through this commit. Amazing work! We'll need to do some pair programming as well next time to better facilitate the transfer of knowledge.

File: src/app/login-button/login-button.component.ts (hunk 3 above)
Comment (A-Robitaille) [reply src/app/login-button/login-button.component.ts]: It was done in the sprint 2 tutorial. Button stuff was being changed anyway.

File: src/app/follow-button/follow-button.component.ts (hunk 5 above)
Comment (tjosepo) [reply src/app/follow-button/follow-button.component.ts:39]: This method is used to detect changes within a component. Angular is supposed to do this automatically whenever a variable changes value, but sometimes it doesn't. I don't think we can use it anywhere else.

**Pull request merged**